drafts/2018-12-01-the-fourth-type-of-variance.md

```diff
@@ -0,0 +1,204 @@
+---
+title: The Fourth Type of Variance
+---
+
+_Variance_ is a way to describe a relationship between different instantiations of the same polymorphic type. Many programmers know about covariance, contravariance, and invariance. Not so many know about the fourth type of variance, or the relationship between them.
+
+
+Covariance
+----------
+
+Haskellers'll be familiar with _covariant functors_. It's the type of functor exhibited the the standard `Functor` class.
+
+```haskell
+class Functor f where
+    fmap :: (a -> b) -> f a -> f b
+```
+
+This chimes with the intuition that a functor is a container of sorts. If you can turn each `a` in the container into a `b`, then you can turn a container of `a`s into a container of `b`s by converting each item in the container. The arrows go in the same direction.
+
+A similar idea is visible in Scala's lattice of subtypes. Type parameters annotated with a `+` are covariant.
+
+```scala
+sealed abstract class List[+A]
+```
+
+A list of `Cat`s is a list of `Animal`s, because every `Cat` is an `Animal`. The subtype relationship of the container goes in the same direction as the subtype relationship of the elements. (In C# `+` is pronounced `out`, as in `IEnumerable<out T>`.)
+
+```scala
+val cats : List[Cat] = List(Cat("Tilly"), Cat("Charlie"), Cat("Izzy"))
+val animals : List[Animal] = cats
+```
+
+Covariant things are producers.
+
+
+Contravariance
+--------------
+
+Less familiar are _contravariant functors_:
+
+```haskell
+class Contravariant f where
+    contramap :: (b -> a) -> f a -> f b
+```
+
+The arrows go in opposite directions. A contravariant functor is a _consumer_ of `a`s. If you can turn `b`s into `a`s, then you can turn a consumer of `a`s into a consumer of `b`s by converting the `b`s into `a`s before they go into the consumer.
```

Comment:
i'm confused how this functor is a consumer of a's -- isn't it taking in Bs and turning them into As so it is producing As and consuming Bs? Once again, this could just be a "me" thing as I'm not familiar with these concepts.